drivers/sensor/nuvoton_tach_npcx/tach_nuvoton_npcx.c

```diff
@@ -0,0 +1,390 @@
+ *                            |---+--->| Counter 2 |<--| _| |_| |_ |<--+
+ *                            |   |    +-----------+   +-----------+
+ *                            | CLK_SEL                Edge Detection
+ *                            |
+ *                            | TACH_CLK
+ *                            +----------
+ *          (NPCX Tachometer Mode 5, Dual-Independent Input Capture)
+ *
+ * This mode is used to measure either the frequency of two external clocks
+ * (via TA or TB pins) that are slower than TACH_CLK. A transition event (rising
+ * or falling edge) received on TAn/TBn pin causes a transfer of timer 1/2
+ * contents to Capture register and reload counter. Based on this value, we can
+ * compute the current RPM of external signal from encoders.
+ *
+ */
+
+#include <errno.h>
+#include <device.h>
+#include <drivers/clock_control.h>
+#include <drivers/sensor.h>
+#include <dt-bindings/sensor/npcx_tach.h>
+#include <soc.h>
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(tach_npcx, CONFIG_SENSOR_LOG_LEVEL);
+
+/* Device config */
+struct tach_npcx_config {
+	/* tachometer controller base address */
+	uintptr_t base;
+	/* clock configuration */
+	struct npcx_clk_cfg clk_cfg;
+	/* pinmux configuration */
+	const uint8_t alts_size;
+	const struct npcx_alt *alts_list;
+	/* sampling clock frequency of tachometer */
+	uint32_t sample_clk;
+	/* selected port of tachometer */
+	int port;
+	/* number of pulses (holes) per round of tachometer's input (encoder) */
+	int pulses_per_round;
+};
+
+/* Driver data */
+struct tach_npcx_data {
+	/* Input clock for tachometers */
+	uint32_t input_clk;
+	/* Captured counts of tachometer */
+	uint32_t capture;
+};
+
+/* Driver convenience defines */
+#define DRV_CONFIG(dev) ((const struct tach_npcx_config *)(dev)->config)
+
+#define DRV_DATA(dev) ((struct tach_npcx_data *)(dev)->data)
+
+#define HAL_INSTANCE(dev) (struct tach_reg *)(DRV_CONFIG(dev)->base)
+
+/* Maximum count of prescaler */
+#define NPCX_TACHO_PRSC_MAX 0xff
+/* Maximum count of counter */
+#define NPCX_TACHO_CNT_MAX 0xffff
+/* Operation mode used for tachometer */
+#define NPCX_TACH_MDSEL 4
+/* Clock selection for tachometer */
+#define NPCX_CLKSEL_APBCLK 1
+#define NPCX_CLKSEL_LFCLK 4
+
+/* TACH inline local functions */
+static inline int tach_npcx_configure(const struct device *dev)
+{
+	const struct tach_npcx_config *const config = DRV_CONFIG(dev);
+	struct tach_npcx_data *const data = DRV_DATA(dev);
+	struct tach_reg *const inst = HAL_INSTANCE(dev);
+
+	/* Set mode 5 to tachometer module */
+	SET_FIELD(inst->TMCTRL, NPCX_TMCTRL_MDSEL_FIELD, NPCX_TACH_MDSEL);
+
+	/* Configure clock module and its frequency of tachometer */
+	if (config->sample_clk == 0) {
+		return -EINVAL;
+	} else if (data->input_clk == LFCLK) {
+		/* Enable low power mode */
+		inst->TCKC |= BIT(NPCX_TCKC_LOW_PWR);
+		if (config->sample_clk != data->input_clk) {
+			LOG_ERR("%s operate freq should be fixed to 32kHz",
+								dev->name);
```

Comment:
Add the APB1 clock speed to the error output.

Reply:
Done.